core/crypto/src/key_file.rs

```diff
@@ -22,10 +57,242 @@ impl KeyFile {
+    key_file.write_to_file(&key_store_path);
+    
+    if home_dir.join("config.json").exists() {
+        //add password corrrectness validation to from_encrypted_file_with_password
+        KeyFile::from_encrypted_file_with_password(&key_store_path)?;
+    } else {
+        key_file.write_to_encrypted_file(&key_store_path)?;
+    }
+
+    Ok(())
+}
+
+pub struct PasswordInput {
+    pub content: Vec<u8>,
+    pub length: usize
+}
+
+impl PasswordInput {
+    // should support that user might not want to encrypt the key
+    pub fn enter_from_console() -> Self {
+        println!("Enter a password for a secret key: ");
+        let mut input = String::new();
+        let _ = std::io::stdin().read_line(&mut input).expect("Could not read from stdin");
+        let content = input.trim().as_bytes().to_vec();
+        let length = input.len();
+        PasswordInput { content, length }
+    }
+
+    pub fn exceed_max_length(&self) -> bool {
+        const MAX_LENGTH: usize = 32;
+        self.length > MAX_LENGTH
+    }
+
+    // xsalsa20poly1305 restricts key size to 32
+    // https://github.com/sodiumoxide/sodiumoxide/blob/master/src/crypto/secretbox/xsalsa20poly1305.rs#L50
+    pub fn pad_content(&self) -> Vec<u8> {
+        let mut buf = [0u8; 32];
+        buf.copy_from_slice(&self.content[..]);
+        buf.to_vec()
+    }
+
+    pub fn not_empty(&self) -> bool {
+        self.length > 0
+    }
+
+    pub fn process() -> Result<PasswordInput, Error> {  
+        let mut password_input = PasswordInput::enter_from_console();
+        if password_input.exceed_max_length() {
+            return Err(Error::new(ErrorKind::InvalidInput, "Input is too long"))
+        }
+       
+        if password_input.not_empty() {
+            let content = password_input.pad_content();
+            let length = content.len();
+            password_input = PasswordInput{content, length};
+        }
+
+        Ok(password_input)
+    }
+
+    pub fn gen_key_from_password_input(&self) -> Result<Key, Error>  {
+        let key = SodiumoxideSecretBox::gen_key(&self.content)?;
+        Ok(key)
+    }
+}
+
+trait Operation {
+    fn encrypt(&self, m: &[u8], k: &Key) -> Vec<u8>;
+    fn decrypt(&self, c: &[u8], k: &Key) -> Vec<u8>;
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+struct SodiumoxideSecretBox {
+    nonce: secretbox::Nonce,
+}
+
+impl SodiumoxideSecretBox {
+    pub fn new() -> Self {
+        let nonce = secretbox::gen_nonce();
+        SodiumoxideSecretBox { nonce }
+    }
+
+    //need to be tested
+    pub fn gen_key(bytes: &[u8]) -> Result<Key, Error> {
+        Key::from_slice(bytes).map(Ok).unwrap_or_else(|| {
+            Err(Error::new(
+                ErrorKind::InvalidInput,
+                "Failed to create key from slice".to_string(),
+            ))
+        })
+    }
+
+}
+
+impl Operation for SodiumoxideSecretBox {
+    fn encrypt(&self, m: &[u8], k: &Key) -> Vec<u8> {
+        secretbox::seal(m, &self.nonce, &k)
```

Comment:
AFAIU secretbox documentation we should be using the same nonce to seal and unseal. We however generate a new  nonce every tiem through `gen_nonce`. I think it was not revealed because there is no test that tests this functionality. We should fix this and have a test that uses a real temporary file (we can have an in-memory file).

Reply:
@nearmax, do you think it is a good idea to add nonce field to KeyFile for it's serialization and enable decryption functionality with the same nonce?

Reply:
Do we have an alternative?